File: 0093-restore-ip-addresses/0093-restore-ip-addresses.py
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        def isValid(s):
            if len(s) > 1 and s[0] == "0":
                return False
            if int(s) < 0 or 255 < int(s):
                return False
            return True

        def dfs(start, path):
            if start == len(s) and len(path) == 4:
                result.append(".".join(path))
                return

            end = start
            while end < start + 3 and start < len(s):
                segment = s[start : end + 1]
                if isValid(segment):
                    path.append(segment)
                    dfs(end + 1, path)
                    path.pop()
                end += 1

        result = []
        dfs(0, [])

        return result

File: 0093-restore-ip-addresses/0093-restore-ip-addresses.js
/**
 * @param {string} s
 * @return {string[]}
 */
var restoreIpAddresses = function (s) {
    const dfs = function (start, path) {
        if (start === s.length && path.length === 4) {
            result.push(path.join("."));
            return;
        }

        for (let end = start; end < start + 3 && start < s.length; end++) {
            let digits = s.substring(start, end + 1);
            if (isValid(digits)) {
                path.push(digits);
                dfs(end + 1, path);
                path.pop();
            }
        }
    }

    const result = [];
    dfs(0, []);

    return result;
};

var isValid = function (s) {
    if (s.length > 1 && s[0] === '0') {
        return false;
    }

    if (Number(s) < 0 || 255 < Number(s)) {
        return false;
    }

    return true;
}